CryptoSwift/Rabbit.swift

```diff
@@ -0,0 +1,206 @@
+    
+    /// 'a' constants
+    private var a: [UInt32] = [
+        0x4D34D34D,
+        0xD34D34D3,
+        0x34D34D34,
+        0x4D34D34D,
+        0xD34D34D3,
+        0x34D34D34,
+        0x4D34D34D,
+        0xD34D34D3,
+    ]
+    
+    // MARK: - Initializers
+    convenience public init?(key:[UInt8]) {
+        self.init(key: key, iv: nil)
+    }
+    
+    public init?(key:[UInt8], iv:[UInt8]?) {
+        self.key = key
+        self.iv = iv
+        
+        guard key.count == Rabbit.keySize && (iv == nil || iv!.count == Rabbit.ivSize) else {
+            return nil
+        }
+    }
+    
+    // MARK: -
+    private func setup() {
+        p7 = 0
+        
+        // Key divided into 8 subkeys
+        var k = [UInt32](count: 8, repeatedValue: 0)
+        for var j = 0; j < 8; j++ {
+            k[j] = UInt32(key[Rabbit.blockSize - (2*j + 1)]) | (UInt32(key[Rabbit.blockSize - (2*j + 2)]) << 8)
+        }
+        
+        // Initialize state and counter variables from subkeys
+        for var j = 0; j < 8; j++ {
+            if j % 2 == 0 {
+                x[j] = (k[(j+1) % 8] << 16) | k[j]
+                c[j] = (k[(j+4) % 8] << 16) | k[(j+5) % 8]
+            } else {
+                x[j] = (k[(j+5) % 8] << 16) | k[(j+4) % 8]
+                c[j] = (k[j] << 16)         | k[(j+1) % 8]
+            }
+        }
+        
+        // Iterate system four times
+        nextState()
+        nextState()
+        nextState()
+        nextState()
+        
+        // Reinitialize counter variables
+        for var j = 0; j < 8; j++ {
+            c[j] = c[j] ^ x[(j+4) % 8]
+        }
+        
+        if let iv = iv {
+            setupIV(iv)
+        }
+    }
+    
+    private func setupIV(iv: [UInt8]) {
+        // 63...56 55...48 47...40 39...32 31...24 23...16 15...8 7...0 IV bits
+        //    0       1       2       3       4       5       6     7   IV bytes in array
+        let iv0: UInt32 = integerWithBytes([iv[4], iv[5], iv[6], iv[7]])
+        let iv1: UInt32 = integerWithBytes([iv[0], iv[1], iv[4], iv[5]])
+        let iv2: UInt32 = integerWithBytes([iv[0], iv[1], iv[2], iv[3]])
+        let iv3: UInt32 = integerWithBytes([iv[2], iv[3], iv[6], iv[7]])
+        
+        // Modify the counter state as function of the IV
+        c[0] = c[0] ^ iv0
+        c[1] = c[1] ^ iv1
+        c[2] = c[2] ^ iv2
+        c[3] = c[3] ^ iv3
+        c[4] = c[4] ^ iv0
+        c[5] = c[5] ^ iv1
+        c[6] = c[6] ^ iv2
+        c[7] = c[7] ^ iv3
+        
+        // Iterate system four times
+        nextState()
+        nextState()
+        nextState()
+        nextState()
+    }
+    
+    private func nextState() {
+        // Before an iteration the counters are incremented
+        var carry = p7
+        for var j = 0; j < 8; j++ {
+            let prev = c[j]
+            c[j] = prev &+ a[j] &+ carry
+            carry = prev > c[j] ? 1 : 0 // detect overflow
+        }
+        p7 = carry // save last carry bit
```

Comment:
I think that's not necessary. 

`p7` is a part of the cipher state which consists of 513 bits (`x[0..7] x 32bit`, `c[0..7] x 32bit` and `p7`) and there is no need to pass it explicitly in `nextState()` method.